include/nbl/builtin/hlsl/subgroup/basic_portability.hlsl

```diff
@@ -0,0 +1,127 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_SUBGROUP_BASIC_PORTABILITY_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SUBGROUP_BASIC_PORTABILITY_INCLUDED_
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace subgroup
+{
+	static const uint MinSubgroupSizeLog2 = 2;
+	static const uint MinSubgroupSize = 0x1u << MinSubgroupSizeLog2;
+	
+#ifdef NBL_HLSL_IMPL_GL_NV_shader_thread_group
+	static const uint MaxSubgroupSizeLog2 = 5;
+#elif defined(NBL_HLSL_IMPL_GL_AMD_gcn_shader)||defined(NBL_HLSL_IMPL_GL_ARB_shader_ballot)
+	static const uint MaxSubgroupSizeLog2 = 6;
+#else
+	static const uint MaxSubgroupSizeLog2 = 7;
+#endif
+	static const uint MaxSubgroupSize = (0x1u << MaxSubgroupSizeLog2);
+	
+	// REVIEW: No need for #ifdef check because these are always available in HLSL 2021 it seems
+	uint Size() {
+		return WaveGetLaneCount();
+	}
+	
+	uint SizeLog2() {
+		return firstbithigh(Size());
+	}
+	
+	uint InvocationID() {
+		return WaveGetLaneIndex();
+	}
+
+	uint SubgroupID() {
+		return gl_LocalInvocationIndex >> SizeLog2();
```

Comment:
A. where do you define `gl_LocalInvocationIndex` ?
B. I think there's a issue open in Vulkan/SPIR-V spec that subgroups don't always fill a workgroup in the same order `gl_LocalInvocationIndex` does

Reply:
you need to get the actual `SubgroupID` with a real SPIR-V intrinsic, see https://stackoverflow.com/questions/69412542/do-subgroup-invocation-indices-map-to-gl-localinvocationindex/69412769#69412769

P.S. I know the old GLSL impl had the same issue.

Reply:
ok we're blocked by https://github.com/microsoft/DirectXShaderCompiler/issues/5455